Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -24,31 +24,68 @@ import PackageLoading
 
 public typealias URL = Foundation.URL
 
+private struct WeakWorkspace {
+  weak var value: Workspace?
+}
+
 /// The SourceKit language server.
 ///
 /// This is the client-facing language server implementation, providing indexing, multiple-toolchain
 /// and cross-language support. Requests may be dispatched to language-specific services or handled
 /// centrally, but this is transparent to the client.
 public final class SourceKitServer: LanguageServer {
-
-  struct LanguageServiceKey: Hashable {
-    var toolchain: String
-    var language: Language
-  }
-
   var options: Options
 
   let toolchainRegistry: ToolchainRegistry
 
-  var languageService: [LanguageServiceKey: ToolchainLanguageServer] = [:]
+  var capabilityRegistry: CapabilityRegistry?
+
+  var languageServices: [ToolchainLanguageServer] = []
 
   /// Documents that are ready for requests and notifications.
   /// This generally means that the `BuildSystem` has notified of us of build settings.
   var documentsReady: Set<DocumentURI> = []
 
   private var documentToPendingQueue: [DocumentURI: DocumentNotificationRequestQueue] = [:]
 
-  public var workspace: Workspace?
+  private let documentManager = DocumentManager()
+
+  /// **Public for testing**
+  public var _documentManager: DocumentManager {
+    return documentManager
+  }
+
+  /// Queue guarding `workspaces` and `uriToWorkspaceCache`.
+  let workspaceQueue = DispatchQueue(label: "", qos: .userInitiated)
```

Comment:
Looking through the code again, I can’t remember any reason why I introduced this queue in the first place instead of just using `self.queue` so I just removed it.